### PR TITLE
API: Deprecated normalized keyword in Rotation

### PR DESCRIPTION
**Pull request by nmayorov**
#### Reference issue

#10968

#### What does this implement/fix?

As explained in #10968 `normalized` keyword was not a good choice. There are serval options to improve the situation, but probably the easiest is to remove this keyword (through deprecation) and always do the normalization ---  it's not expensive operation and it guarantees that the `Rotation` has the valid state.

Note that in `Rotation.__init__` I changed `normalized` to `normalize` with the opposite meaning. But it is stated in the documentation that `__init__` is not supposed to used directly, so I think this is a legit change.

I think it would be good to make this change for 1.4.0, if do it at all.

@larsoner @pmla @rgommers

### CODE REVIEW

**Comment by larsoner** [comment]
@pmla can you look and see if you're happy? If so I can take a look and merge.

Would be good to get it in within the next couple of days to be safe for 1.4.0 branching.

File: scipy/spatial/transform/rotation.py [hunk 1]
```diff
@@ -1864,9 +1855,13 @@ def match_vectors(cls, a, b, weights=None, normalized=False):
                                     weights.shape[0], b.shape[0]))
         weights = weights / np.sum(weights)
 
-        if not normalized:
-            a = a / scipy.linalg.norm(a, axis=1)[:, None]
-            b = b / scipy.linalg.norm(b, axis=1)[:, None]
+        if normalized is not None:
+            warnings.warn("`normalized` is deprecated in scipy 1.4.0 and "
+                          "will be removed in scipy 1.6.0. The input vectors "
+                          "are always normalized.", DeprecationWarning)
```
Comment (pmla) [inline scipy/spatial/transform/rotation.py]: This should be "Normalization is not performed" or similar.

File: scipy/spatial/transform/rotation.py [hunk 2]
```diff
@@ -1864,9 +1855,13 @@ def match_vectors(cls, a, b, weights=None, normalized=False):
...
+
+        a = a / scipy.linalg.norm(a, axis=1)[:, None]
+        b = b / scipy.linalg.norm(b, axis=1)[:, None]
```
Comment (pmla) [inline scipy/spatial/transform/rotation.py]: Normalization should not be performed here.

**Comment by pmla** [comment]
I'm happy with the API. I have added a couple of comments about the code.

File: scipy/spatial/transform/rotation.py [hunk 3]
```diff
@@ -1801,36 +1813,26 @@ def match_vectors(cls, a, b, weights=None, normalized=False):
             Vector components observed in another frame B. Each row of `b`
             denotes a vector.
         weights : array_like shape (N,), optional
-            Weights describing the relative importance of the vectors in
-            `a`. If None (default), then all values in `weights` are assumed to
-            be equal.
-        normalized : bool, optional
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I suppose I must retain `normalized` description for now and explain that it is deprecated (rather just completely hide it)?

@rgommers is it a right policy here?

Comment (rgommers) [reply scipy/spatial/transform/rotation.py]: Yes that would be the normal way of doing that.

File: scipy/spatial/transform/rotation.py [hunk 4]
```diff
@@ -365,7 +365,7 @@ class Rotation(object):
     output formats supported, consult the individual method's examples.
 
     """
-    def __init__(self, quat, normalized=False, copy=True):
+    def __init__(self, quat, normalize=True, copy=True):
```
Comment (rgommers) [inline scipy/spatial/transform/rotation.py]: here you're renaming a keyword, that's a backwards compat break right? or was this not in 1.3.0?

**Review by rgommers — COMMENTED** [review]
I think the PR description and code changes have gotten out of sync, would be good to update the description. 

I don't understand the changes from a quick read - `normalized` is deprecated, but it's also replaced with `normalize` ?

File: scipy/spatial/transform/rotation.py (hunk 4 above)
Comment (nmayorov) [reply scipy/spatial/transform/rotation.py]: The situation here is slightly unclear. The class was designed as such as not to use `__init__` directly and this was clearly stated in the documentation from the very beginning. 

So I suppose we can make this change without violating any "contracts". Do you agree?

Comment (rgommers) [reply scipy/spatial/transform/rotation.py]: Ah, that makes sense, rings a bell. Okay, then yes I agree.

**Comment by larsoner** [comment]
@pmla happy now? If so it looks like @rgommers is also satisfied / comments have been addressed so I'll merge

**Comment by nmayorov** [comment]
No, wait. I need to restore descriptions of deprecated parameters.

---- On Wed, 13 Nov 2019 20:07:41 +0500 notifications@github.com wrote ----



@pmla happy now? If so it looks like @rgommers is also satisfied / comments have been addressed so I'll merge

**Comment by larsoner** [comment]
@nmayorov let me know when it's done. 1.4.0 branches in the next couple of days.

File: scipy/spatial/transform/rotation.py (hunk 3 above)
Comment (nmayorov) [reply scipy/spatial/transform/rotation.py]: Now I see a problem here. We want to change the default behavior of `match_vectors` with respect to `normalized` keyword. So it's going to be a backward incompatible change at some point.

Eventually we want to remove `normalized` and never do the normalization, but now it is done by default.

Conceptually it looks like a truly backward incompatible change. Should we simply introduce it now? I don't see any other way.

File: scipy/spatial/transform/rotation.py [hunk 5]
```diff
@@ -1864,9 +1866,14 @@ def match_vectors(cls, a, b, weights=None, normalized=False):
                                     weights.shape[0], b.shape[0]))
         weights = weights / np.sum(weights)
 
-        if not normalized:
-            a = a / scipy.linalg.norm(a, axis=1)[:, None]
-            b = b / scipy.linalg.norm(b, axis=1)[:, None]
+        if normalized is not None:
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: This whole block is broken, need to make a decision about the change as I explained above.

**Comment by nmayorov** [comment]
@rgommers @larsoner 

I did remove `normalized` parameter from `Rotation.math_vectors` in a backward incompatible way, i.e. default behavior is changed, the argument is completely removed.

As I understand the "correct" way to do that is to 1) Advertise the change of default behavior through `FutureWarning` 2) Do the change 3) Advertise deprecation 4) Do the deprecation. I am not exactly sure of this cycle still.

Instead of going through this quite convoluted cycle, I suggest we just announce this change in the release notes and move on.

Happy to hear better suggestions.

**Comment by tylerjereddy** [comment]
I find this all very confusing... I'm tempted to delay the merge and just suggest a clarification at the docs level for some time. If some kind of deprecation cycle is going to happen eventually is it really that critical to rush it?

**Comment by rgommers** [comment]
I'll have a look at this now. Since if we do what @nmayorov suggests, it has to be done in 1.4.0

**Comment by rgommers** [comment]
Okay, I see what the issue is @nmayorov is running into. The `FutureWarning` thing is tricky to do in this case, because there's no good way to silence that warning or to prepare for the coming change. Doing that via a temporary keyword or via the keyword we're about to remove is a bad idea, because then that keyword removal is a backwards-incompatible change anyway. And users will have to change their code twice.

It's not nice, because we're changing the returned numerical values here for input vectors that are not normalized and not of equal length. But short of renaming the method, that is the best option it looks like.

**Comment by rgommers** [comment]
Okay, merged (and squashed, that seemed better to have a clean diff rather than the changes of direction in the history). 

I'm okay to go with @nmayorov's proposal here. @larsoner and @pmla please have another good look at this. If you think that there's a better alternative we can make them before the final release.

**Comment by rgommers** [comment]
I'll write an addition to the release notes now.

**Comment by rgommers** [comment]
@tylerjereddy could you add this to the release notes?
```
`scipy.spatial` changes
-----------------------
The default behavior of the ``match_vectors`` method of `scipy.spatial.transform.Rotation`
was changed for input vectors that are not normalized and not of equal lengths.
Previously, such vectors would be normalized within the method.  Now, the calculated
rotation takes the vector length into account, longer vectors will have a larger
weight. For more details, see https://github.com/scipy/scipy/issues/10968.
```

**Comment by tylerjereddy** [comment]
Yeah, alright

**Comment by pmla** [comment]
How about we deprecation warning for the whole function and write a new one?
The current function name is unfortunate: `match_vectors` implies that a matching will be calculated. What this function does is to _align_ vectors. `align_vectors` would be better.

I think this is more than a nitpick. For example, the aim of the 3D registration problem is to compute the optimal permutation (or "matching") and the optimal rotation. The term "matching" means something very specific in the mathematical literature, and we also use it this way in other parts of scipy (see e.g. `scipy.sparse.csgraph.maximum_bipartite_matching` or the description of `scipy.optimize.linear_sum_assignment`).

The other thing we could change at the same time is the return value. The alignment problem finds the optimal rotation under the root-mean-square distance RMSD, yet this value is not returned.

**Comment by nmayorov** [comment]
> How about we deprecation warning for the whole function and write a new one?
> The current function name is unfortunate: `match_vectors` implies that a matching will be calculated. What this function does is to _align_ vectors. `align_vectors` would be better.
> 
> I think this is more than a nitpick. For example, the aim of the 3D registration problem is to compute the optimal permutation (or "matching") and the optimal rotation. The term "matching" means something very specific in the mathematical literature, and we also use it this way in other parts of scipy (see e.g. `scipy.sparse.csgraph.maximum_bipartite_matching` or the description of `scipy.optimize.linear_sum_assignment`).
> 
> The other thing we could change at the same time is the return value. The alignment problem finds the optimal rotation under the root-mean-square distance RMSD, yet this value is not returned.

Sounds very reasonable.

@tylerjereddy I can work on this later today. We can put it into 1.4.0, right?

**Comment by tylerjereddy** [comment]
> I can work on this later today. We can put it into 1.4.0, right?

Yep, I think Ralf overruled me on putting this stuff in 1.4.0. It looks like he is in favor of your second suggestion there too. Sorry I can't parse all the details right now, but I trust you and Peter seem to be thinking about it carefully.

Adding small release note blurbs for me to paste in may be helpful.

**Pull request merged**